Generating Enum Classes

FlaminGuy opened the pull request:
Class generation for enums.
At this stage we only emit classes for enums. These classes are not in use at this point.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala:
```diff
@@ -0,0 +1,353 @@
+    // Generate `equals` method
+    compileEqualsMethod(visitor, name, tpe, fType)
+
+    visitor.visitEnd()
```
I am really happy that you separated out the methods compileToStringMethod etc etc.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala:
```diff
@@ -0,0 +1,353 @@
+    visitor.toByteArray
+  }
+
+  /**
```
It might be useful to give en example as part of the comment.

E.g. given an enum of this structure (you can use list) then this method generates the following class (shown in Java code)

Same goes for every other method.

FlaminGuy replied:
I added examples.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala:
```diff
@@ -0,0 +1,353 @@
+    // Generate the `getValue` method
+    // Picking the right opcode for return
+    val retCode = fType match {
+      case Type.Var(id, kind) =>  throw InternalCompilerException(s"Non-monomorphed type variable '$id in type '$tpe'.")
```
Correct :)

FlaminGuy replied:
Corrected :)

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala:
```diff
@@ -0,0 +1,353 @@
+      case _ => ARETURN
+    }
+    compileGetFieldMethod(visitor, name, tpe, "value", "getValue", retCode)
+    // Generate the `getTag` method
```
Add line break.

FlaminGuy replied:
Done!

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala:
```diff
@@ -0,0 +1,353 @@
+    * @param fType Type of the `value` field
+    * @return byte code representation of the class
+    */
+  def compileEnumClass(name: List[String], enumInterface: Class[_], tpe: String, fType: Type): Array[Byte] = {
```
Again, avoid use of `Class`.

FlaminGuy replied:
Sure thing. Done!

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala:
```diff
@@ -0,0 +1,353 @@
+  }
+
+  /**
+    * Creates constructor of the enum case class which is named `name`.
```
Again, show an example of the generated code in pseudo-Java code.

FlaminGuy replied:
Added examples.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala:
```diff
@@ -0,0 +1,353 @@
+    * @param fType type of the `value` field
+    */
+  private def compileEnumConstructor(visitor: ClassWriter, name: List[String], tpe: String, fType: Type) = {
+    val constructor = visitor.visitMethod(ACC_PUBLIC, "<init>", s"($tpe)V", null, null)
```
// Generate single argument constructor (right??)

FlaminGuy replied:
Right!

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala:
```diff
@@ -0,0 +1,353 @@
+    constructor.visitVarInsn(iLoad, 1)
+    constructor.visitFieldInsn(PUTFIELD, decorate(name), "value", tpe)
+
+
```
extra linebreak

FlaminGuy replied:
Done!

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala:
```diff
@@ -0,0 +1,353 @@
+
+    method.visitInsn(IRETURN)
+    method.visitMaxs(1, 1)
+    method.visitEnd()
```
Looks good to me.

FlaminGuy replied:
Same.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala:
```diff
@@ -0,0 +1,353 @@
+        method.visitJumpInsn(IFNE, neq)
+        method.visitInsn(ICONST_1)
+      case _ =>
+        method.visitMethodInsn(INVOKEVIRTUAL, asm.Type.getInternalName(clazz), objectEqualsMethod.getName,
```
This is where we have to be careful with tuples. At least until we get the code gen in place for tuples. Maybe add a //TODO

FlaminGuy replied:
I think this code is correct when we generate tuples because we have to generate `equals()` but it won't work now. So perhaps, there is nothing `TODO` now or in future? :)

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/LoadBytecode.scala:
```diff
@@ -63,21 +63,43 @@ object LoadBytecode extends Phase[ExecutableAst.Root, ExecutableAst.Root] {
     */
   def run(root: ExecutableAst.Root)(implicit flix: Flix): Validation[ExecutableAst.Root, CompilationError] = {
     implicit val _ = flix.genSym
+    val fixedPrefix = List("ca", "waterloo", "flix", "enums")
```
I think we should put them in some other namespaces, but let us discuss tomorrow. It might even be possible to use the Flix namespace as the JVM namespace.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/LoadBytecode.scala:
```diff
@@ -105,16 +127,81 @@ object LoadBytecode extends Phase[ExecutableAst.Root, ExecutableAst.Root] {
+    // 4. Extract declared enums.
+    val declaredEnums: Map[Type, Set[(String, Type)]] = root.definitions.values
+      .flatMap(x => findEnums(x.exp)).groupBy(x => x._1)
+      .mapValues(x => x.map(y => y._2).toSet)
```
mapValues is super dangerous because it is lazy. I will let ifaz explain why tomorrow :)

magnus-madsen replied:
Basically the Scala people screwed up when they implemented mapValues.

FlaminGuy replied:
I knew about this, but I think since I immediately use declaredEnums I force the lazy val to get evaluated.
I got rid of this so we should be good either way.

magnus-madsen replied:
Ok; great. We have had tons of bugs with subtle uses of `mapValues` so now I am just totally banning it from Flix !

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/LoadBytecode.scala:
```diff
@@ -105,16 +127,81 @@ object LoadBytecode extends Phase[ExecutableAst.Root, ExecutableAst.Root] {
+      val prefix : List[String] = fixedPrefix ++ sym.namespace ++ typeToUniquePath(tpe)
+      // Full name
+      val fullName : List[String] = prefix ++ List("EnumInterface")
+      val bytecode = Enumgen.compileEnumInterface(fullName, tagInterface)
```
Notice how you depend on tagInterface, but this really should not be necessary.

FlaminGuy replied:
Done!

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/LoadBytecode.scala:
```diff
@@ -233,4 +320,67 @@ object LoadBytecode extends Phase[ExecutableAst.Root, ExecutableAst.Root] {
+    * @param tpe type
+    * @return
+    */
+  private def typeToUniquePath(tpe: Type) : List[String] = {
```
Let us discuss tomorrow. I would prefer a translation that uses a single character, e.g. $.

FlaminGuy replied:
I got ride of this, we either use the name of primitive or `object` after the namespace and symbol so the names will be unique.

magnus-madsen replied:
Example?

The user is allowed to write evil stuff like:

```
enum ListInt {}
// or 
enum Foo {
  case ListInt
}
```

so we have to be careful. I would expect that the only safe way to solve this is to use some special  symbol.

FlaminGuy replied:
So if you look now we don't use it.
For the following enum:
` enum Option[Int32] {
    case None,
    case Some(Int32)
}` 

we create `ca.waterloo.flix.enums.Option.Int32.Some` and `ca.waterloo.flix.enums.Option.object.None`.
For 
` enum Option[List[Int32]] {
    case None,
    case Some(List[Int32])
}` 
We create ca.waterloo.flix.enums.Option.object.Some` and we will use the same `ca.waterloo.flix.enums.Option.object.None`. Naming is unique this way.

magnus-madsen replied:
What about?

```
namespace Int32 {
  enum Some
}
```

The question is whether we feel 100% safe. I will think about it.

FlaminGuy replied:
We will create `ca.waterloo.flix.enums.Int32.object.some`.
The general formatting is `ca.waterloo.flix.enums.NAMESPACE.ENUMNAME.TYPEOFFIELD.CASENAME`. `ca.waterloo.flix.enums` is to differentiate generated codes from other codes, `NAMESPACE.ENUMNAME` should uniquely identify the enum. `TYPEOFFIELD` should uniquely identify type of the value field and `CASENAME` must be unique too.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/LoadBytecode.scala:
```diff
@@ -233,4 +320,67 @@ object LoadBytecode extends Phase[ExecutableAst.Root, ExecutableAst.Root] {
+    * @param e expression
+    * @return
+    */
+  def findEnums(e: Expression): List[(Type, (String, Type))] = e match {
```
It should be possible to move this into EnumGen (at least after there is no dependencies on Class[_])

FlaminGuy replied:
We still need this for CodeGen phase but I moved this to CodegenHelper.

magnus-madsen commented:
Great job! The code structure is good and the code is clear. It is only numerous minor issues.

The three biggest issues that we have to fix is:

- the dependence on Class[_]
- the use of `List[String]` etc. when we should really be passing around proper "AST fragments" (e.g. symbols and possibly some types). 
- A proper mapping from types to JVM names.

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala:
```diff
@@ -0,0 +1,353 @@
+    * @param name name of the class
+    * @return byte code representation of the class
+    */
+  def compileTagInterface(name: List[String]) : Array[Byte] = {
```
Passing a `List[String]` around as the name of the class is not only in my code. It is used everywhere. Also, you have to make this conversion at more places if you don't do it in earlier steps. I'm actually refactoring all the places that we use `List[String]` and I'm using a case class called `QualName(ref: List[String])`. `QualName` comes from `QualifiedName` which is basically address of classes in Java. If you don't like this convention then again we can change it, but take a look at the code which I'll push shortly to see if this is an acceptable solution.

magnus-madsen replied:
I much prefer a case class like `QualName` since that at least carries some semantic information. Perhaps I might suggest classes like:

`JavaClassName`, `JavaInterfaceName`, etc. etc. if that makes sense?

I understand that some of the older code uses strings - but I am not satisfied with that either :)

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala:
```diff
@@ -0,0 +1,353 @@
+    * @return byte code representation of the class
+    */
+  def compileTagInterface(name: List[String]) : Array[Byte] = {
+    val visitor = new ClassWriter(0)
```
What do you mean by that. Like adding it as a field?

FlaminGuy replied:
Uh, let's do it when we want to fix the #402 . There is already so many things inside this one pull request.

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala:
```diff
@@ -0,0 +1,117 @@
+import org.objectweb.asm.{Type => _, _}
+
+/**
+  * Created by ramin on 2017-05-22.
```
Done!

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala:
```diff
@@ -0,0 +1,117 @@
+  // This constant is used in LoadBytecode, so we can't put it in the private Constants object.
+  val flixObject = "flixObject"
+
+  case class Context(prefix: List[String],
```
I got rid of it for you.

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala:
```diff
@@ -0,0 +1,353 @@
+import ca.uwaterloo.flix.util.InternalCompilerException
+
+/**
+  * Created by ramin on 2017-05-22.
```
Done!

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala:
```diff
@@ -0,0 +1,353 @@
+  *
+  * Generates classes and interfaces which will represent enums
+  */
+object Enumgen {
```
I made both EnumGen and CodeGen separate phases from LoadBytecode.

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala:
```diff
@@ -0,0 +1,353 @@
+  def compileTagInterface(name: List[String]) : Array[Byte] = {
+    val visitor = new ClassWriter(0)
+
+    visitor.visit(V1_8, ACC_PUBLIC + ACC_ABSTRACT + ACC_INTERFACE, decorate(name), null,
```
I made this a field on flix object which we pass around in different phases.

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala:
```diff
@@ -0,0 +1,353 @@
+    * @param tagInterface the tagInterface
+    * @return byte code representation of the class
+    */
+  def compileEnumInterface(name: List[String], tagInterface: Class[_]): Array[Byte] = {
```
See the comment on the other place.

FlaminGuy replied:
Done!

FlaminGuy commented:
So I fixed almost everything you wanted from me. When I unified all enum classes with an object field to one class I introduced couple of bugs which took me some time to figure out.
Now, there are 34 tests failing. I added a hack in `Value.scala` to get around most of the calls to interpreter from code gen but if you get rid of the hack then 500 more tests will fail. We can get rid of my hack after the call into interpreter is gone. Around 20 tests of the 34 tests failing are because that those tests expect the old `Tag` class which I haven't changed but will do soon. So I'll start on Tuple generation and mainwhile you can fix the dependency on `Interpreter`.

magnus-madsen commented:
I see some changes, but not all the changes you mentioned. Did you push? or did GitHub not update the diff or something?

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala:
```diff
@@ -0,0 +1,650 @@
+    }
+
+    // 1. Transform non-functions and generate interface names.
+    val constantsList: List[Definition.Constant] = root.definitions.values.map { f =>
```
!!!: You are not supposed to change the AST. What is the reason to change the type?

(NB: The types in Code Generator have been a source of bugs/confusion so it is entirely possible that what you are doing makes good sense, but in that case the transformation should probably happen earlier in the compiler),

FlaminGuy replied:
I removed this!

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala:
```diff
@@ -0,0 +1,650 @@
+      }
+    }.toList.filterNot(_.ann.isLaw)
+
+    val interfaceNames: Map[Type, QualName] = generateInterfaceNames(constantsList)
```
Uh, what exactly are these interfaces?

magnus-madsen replied:
My suspicion is that these names should *not* be generated in this phase.

FlaminGuy replied:
removed! this is not needed.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala:
```diff
@@ -0,0 +1,650 @@
+    val visitor = new ClassWriter(0)
+
+    visitor.visit(flix.javaVersion, ACC_PUBLIC + ACC_ABSTRACT + ACC_INTERFACE, decorate(qualName), null,
+      asm.Type.getInternalName(Constants.objectClass), Array(asm.Type.getInternalName(Constants.tagInterface)))
```
Could we extract the last two arguments into some local variables with informative names? I am thinking one of these is the superclass.

FlaminGuy replied:
Done!

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala:
```diff
@@ -0,0 +1,650 @@
+                       interfaceName: QualName,
+                       fType: Option[Type],
+                       interfaces: Map[Type, QualName],
+                       javaVersion: Int): Array[Byte] = {
```
We don't have to pass the javaVersion since we are not going to generate different bytecode in the same run of the compiler. We can just access the javaVersion from some appropriate field.

FlaminGuy replied:
Done!

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala:
```diff
@@ -0,0 +1,650 @@
+    * @return byte code representation of the class
+    */
+  def compileEnumClass(qualName: QualName,
+                       interfaceName: QualName,
```
May I suggest the names: `qualName` -> `className`, `interfaceName` -> `superType`.

FlaminGuy replied:
Done!

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala:
```diff
@@ -0,0 +1,650 @@
+  def compileEnumClass(qualName: QualName,
+                       interfaceName: QualName,
+                       fType: Option[Type],
+                       interfaces: Map[Type, QualName],
```
This map should not really be necessary here. Instead it should be computed when needed.

We can leave this as as separate refactoring.

FlaminGuy replied:
removed! not needed at all.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala:
```diff
@@ -0,0 +1,650 @@
+    val visitor = new ClassWriter(ClassWriter.COMPUTE_FRAMES)
+
+    // Initialize the visitor to create a class.
+    visitor.visit(javaVersion, ACC_PUBLIC + ACC_FINAL, decorate(qualName), null, asm.Type.getInternalName(Constants.objectClass),
```
Let us extract the components out into local variables like `superName` and `interfaces` or similar.

FlaminGuy replied:
Done!

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala:
```diff
@@ -0,0 +1,650 @@
+      Array(decorate(interfaceName.ref)))
+
+    // Descriptor of the the `value` field of the class
+    val fieldDescriptor = getFieldTypeDescriptor(fType, interfaces)
```
I understand that you need `interfaces` here, but it would be nice if the name could be computed purely from the `fType`, but we can leave it for a future refactoring.

FlaminGuy replied:
I changed it now. No need for fieldDescriptor anymore

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala:
```diff
@@ -0,0 +1,650 @@
+
+    // Generate the `getValue` method
+    // Picking the appropriate opcode for return
+    val retCode = fType match {
```
Let us extract this into a helper function since it seems to come up again and again.

FlaminGuy replied:
Done!

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala:
```diff
@@ -0,0 +1,650 @@
+    // Generate `equals` method
+    compileEqualsMethod(visitor, qualName, fieldDescriptor, fType)
+
+    // Generate `getBoxedValue` method
```
Move up after getValue method.

FlaminGuy replied:
Done!

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala:
```diff
@@ -0,0 +1,650 @@
+    *
+    * And for `Error[List[Int32]]` we generate:
+    *
+    * public boolean value;
```
public Object value?

FlaminGuy replied:
Fixed!

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala:
```diff
@@ -0,0 +1,650 @@
+    * @param descriptor descriptor of the `value` field
+    * @param fType type of the `value` field
+    */
+  private def compileEnumConstructor(visitor: ClassWriter, qualName: QualName, descriptor: String, fType: Option[Type]) = {
```
`qualName` -> `className`?

FlaminGuy replied:
Done!

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala:
```diff
@@ -0,0 +1,650 @@
+    val clazz = Constants.objectClass
+    val ctor = clazz.getConstructor()
+
+    constructor.visitCode()
```
Describe what these two lines do

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala:
```diff
@@ -0,0 +1,650 @@
+      asm.Type.getConstructorDescriptor(ctor), false)
+
+    // Load instruction for type of `value`
+    val iLoad : Int = fType match {
```
Probably useful to extract into a helper function, as above.

FlaminGuy replied:
Done!

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala:
```diff
@@ -0,0 +1,650 @@
+      case _ => ALOAD
+    }
+
+    // Put the object given to the construct on the `value` field
```
construct -> constructor

FlaminGuy replied:
Fixed!

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala:
```diff
@@ -0,0 +1,650 @@
+
+    // Put the name of the tag on the `tag` field
+    constructor.visitVarInsn(ALOAD, 0)
+    constructor.visitLdcInsn(qualName.ref.last)
```
We talked about the problem of relying on `.last`, hopefully something can be done.

FlaminGuy replied:
Fixed.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala:
```diff
@@ -0,0 +1,650 @@
+    * @param methodName method name of getter of `fieldName`
+    * @param iReturn opcode for returning the value of the field
+    */
+  private def compileGetFieldMethod(visitor: ClassWriter, qualName: QualName, descriptor: String, fieldName: String,
```
If we create a helper method then we can avoid passing iReturn.

FlaminGuy replied:
Fixed!

magnus-madsen commented:
Lots of comments, but mostly just minor nitpicking things. I only saw bigger issues indicated by !!!.

Otherwise, I think the code is shaping up nicely and starting to look very solid.

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala:
```diff
@@ -0,0 +1,650 @@
+    *
+    * package ca.waterloo.flix.enums.Result;
+    * import ca.uwaterloo.flix.api.TagInterface;
+    * public interface EnumInterface extends TagInterface {
```
No it's called EnumInterface. My reasoning was to make Result be the path that contains the cases and the interface so virtually, in the folder Result we have TagInterface, Ok and Err

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/ast/ExecutableAst.scala:
```diff
@@ -33,6 +35,10 @@ object ExecutableAst {
+                  enumInterfaceByteCodes: Map[EnumSym, (QualName, Array[Byte])],
+                  enumClassByteCodes: Map[EnumSym, (Map[(String, Type), Array[Byte]], Map[String, Array[Byte]])],
+                  interfaceByteCodes: Map[Type, (QualName, Array[Byte])],
+                  classByteCodes: Map[QualName, Array[Byte]],
```
Fixed!

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/ast/Type.scala:
```diff
@@ -95,6 +95,15 @@ sealed trait Type {
+    case Type.Char | Type.Bool | Type.Int8 | Type.Int16 | Type.Int32 | Type.Int64 | Type.Float32 |
+         Type.Float64 => true
+    case _ => false
+  }
```
Fixed!

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala:
```diff
@@ -496,46 +530,82 @@ object Codegen {
+       * if it is primitive then there is no need for casting
+       */
+      tpe match {
+        case Type.Var(id, kind) =>  throw InternalCompilerException(s"Non-monomorphed type variable '$id in type '$tpe'.")
```
Done!

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala:
```diff
@@ -496,46 +530,82 @@ object Codegen {
+          }
+          val fullName = CodegenHelper.getEnumInterfaceName(sym)
+          visitor.visitTypeInsn(CHECKCAST, decorate(fullName))
+        case t if t.isPrimitive =>
```
Fixed!

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala:
```diff
@@ -0,0 +1,261 @@
+
+  val fixedPrefix = List("ca", "waterloo", "flix", "enums")
+
+  case class QualName(ref: List[String])
```
Done!

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala:
```diff
@@ -0,0 +1,261 @@
+
+  case class QualName(ref: List[String])
+
+  def getEnumCaseName(sym: EnumSym, tag: String, tpe: Option[Type]) : QualName = tpe match {
```
Added!

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala:
```diff
@@ -0,0 +1,261 @@
+  case class QualName(ref: List[String])
+
+  def getEnumCaseName(sym: EnumSym, tag: String, tpe: Option[Type]) : QualName = tpe match {
+    case Some(t) if t.isPrimitive => QualName(CodegenHelper.fixedPrefix ++ sym.namespace ++ List(sym.name, t.toString, tag))
```
I fixed both!

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala:
```diff
@@ -0,0 +1,261 @@
+    case _ => QualName(CodegenHelper.fixedPrefix ++ sym.namespace ++ List(sym.name, "object", tag))
+  }
+
+  def getEnumInterfaceName(sym: EnumSym): QualName = {
```
Added!

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala:
```diff
@@ -0,0 +1,261 @@
+  /*
+   * Decorate (mangle) a prefix (list of strings) to get the internal JVM name.
+   */
+  def decorate(prefix: List[String]): String = prefix.mkString("/")
```
Reduced to 1!

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala:
```diff
@@ -0,0 +1,261 @@
+   */
+  def decorate(prefix: List[String]): String = prefix.mkString("/")
+
+  def decorate(qualName: QualName): String = decorate(qualName.ref)
```
Done!

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala:
```diff
@@ -0,0 +1,261 @@
+  def decorate(qualName: QualName): String = decorate(qualName.ref)
+
+  /*
+   * Returns the internal name of the JVM type that `tpe` maps to.
```
The comment still applies, I didn't change the function, I just moved a variable in the case class to be directly given to the function. I changed the format of the comment though so it looks like the rest.

FlaminGuy commented on main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala:
```diff
@@ -0,0 +1,261 @@
+
+
+  /**
+    * Find enums from the given expression
```
Added!

FlaminGuy commented on main/src/ca/uwaterloo/flix/api/Flix.scala:
```diff
@@ -89,6 +89,11 @@ class Flix {
   /**
+    * Java version of generated codes.
+    */
+  val javaVersion = V1_8
```
Fixed!

FlaminGuy commented:
Except two minor comments, all other are fixed and Tuple generation has also been added to this pull.
Testbackend still requires refactoring before pushing which can be done in couple of hours today.

FlaminGuy commented:
Since we were close to remove dependency on Value object, I just removed it completely. 
This pull request should fix #420 #440 #415 #404 #403

magnus-madsen commented:
If you can set the failing hook tests to `@ignore` that would be great. Then all tests should pass.

FlaminGuy commented:
The benchmark result is this:
Before change:
      Name:                  Time (ms)          Ops (ops/s)
      benchmark01                199.9                  5.0
      benchmark02                254.7                  3.9
      benchmark03                233.8                  4.3
      benchmark04                535.2                  1.9
      benchmark05                223.4                  4.5
      benchmark06                155.7                  6.4
      benchmark07                107.1                  9.3
      benchmark08                 98.4                 10.2
      benchmark09                448.4                  2.2

After change:
      Name:                  Time (ms)          Ops (ops/s)
      benchmark01                 94.8                 10.6
      benchmark02                142.3                  7.0
      benchmark03                 84.2                 11.9
      benchmark04                234.9                  4.3
      benchmark05                187.6                  5.3
      benchmark06                 77.4                 12.9
      benchmark07                 57.5                 17.4
      benchmark08                 73.6                 13.6
      benchmark09                308.0                  3.2

The numbers still tolerate a lot but it's clear that we are considerably faster now. 
We are still slower than scala though:
      Name:                  Time (ms)   
      benchmark01                 45.5
      benchmark02                51.2
      benchmark03                 54.0
      benchmark04                104.8
      benchmark05                99.5
      benchmark06                 45.1
      benchmark07                 59.4
      benchmark08                 65.5
      benchmark09                93.0

magnus-madsen commented:
Looks good. Can you paste the Scala code?

magnus-madsen commented on main/src/ca/uwaterloo/flix/api/Flix.scala:
```diff
@@ -26,7 +26,7 @@ import ca.uwaterloo.flix.runtime.quickchecker.QuickChecker
 import ca.uwaterloo.flix.runtime.{DeltaSolver, Model, Solver, Value}
 import ca.uwaterloo.flix.util.{LocalResource, Options, Validation}
-
+import org.objectweb.asm.Opcodes.V1_8
```
Unused import

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala:
```diff
@@ -1256,4 +1366,38 @@ object Codegen {
+   * `tpe` is type of value on top of the stack. If the value is not primitive, then we cast it to it's specific type,
+   * if the value is a primitive then since there is no boxing, then no casting is necessary.
+   */
+  private def castIfNotPrim(tpe: Type, interfaces: Map[Type, FlixClassName], visitor: MethodVisitor) = tpe match {
```
Return type? Unit?

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala:
```diff
@@ -0,0 +1,627 @@
+    val tupleClass : Class[_] = classOf[TupleInterface]
+    val scalaPredef = "scala/Predef$"
+    val scalaMathPkg = "scala/math/package$"
+    val tagInterface : Class[_] = classOf[TagInterface]
```
both tagInterface and tagClass?

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala:
```diff
@@ -0,0 +1,627 @@
+    * @param e expression
+    * @return List of enum cases in the expression, each element of the list is of form (EnumType, (CaseName, EnumFieldType))
+    */
+  def findEnumCases(e: Expression): List[(Type, (String, Type))] = e match {
```
This can go in the EnumGen phase.

FlaminGuy replied:
This can't go there, it is needed in CodeGen to create a map from enum types to QualName.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala:
```diff
@@ -0,0 +1,627 @@
+    * @param e expression
+    * @return A list containing all the tuples in the given expression
+    */
+  def findTuples(e: Expression): List[Type] = {
```
This can go in the TupleGen

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/TreeShaker.scala:
```diff
@@ -43,7 +43,7 @@ object TreeShaker extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
     */
   def run(root: SimplifiedAst.Root)(implicit flix: Flix): Validation[SimplifiedAst.Root, CompilationError] = {
-
+    return root.toSuccess
```
Remove, this should not be here.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/TupleGen.scala:
```diff
@@ -0,0 +1,464 @@
+import org.objectweb.asm
+import org.objectweb.asm.{ClassWriter, Label}
+import org.objectweb.asm.Opcodes._
+object TupleGen extends Phase[ExecutableAst.Root, ExecutableAst.Root]{
```
line break

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/TupleGen.scala:
```diff
@@ -0,0 +1,464 @@
+
+  /**
+    * Generate bytecode for Tuples.
+    * The steps that we take to generate enums is as follows:
```
tuples

magnus-madsen commented:
I tried to limit my comments so we can this merged in. 
Then we can work on improvements in other tickets. 

Once the few last things are fixed and the tests all pass (with the hooks set to `ignore`) then I am happy to merge in.

FlaminGuy commented:
I fixed those things, cleaned a few things left. 
We should be good to go.

FlaminGuy commented:
Please close #440 #420 #415 #404 #403 after merging this in @magnus-madsen.

magnus-madsen commented:
Thanks for your great work on this :)!

I am happy to see that our changes have a measurable impact on performance and that we are moving closer to Scala in terms of execution time.

There are still things to be done, but we will fix them in smaller tickets/pull requests as we go along.

Again, thanks for great work :)

FlaminGuy commented:
No problem Magnus, I enjoyed working on this 👍
